pset/blinder.go

```diff
@@ -0,0 +1,314 @@
+package pset
+
+import (
+	"crypto/rand"
+	"errors"
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/vulpemventures/go-elements/confidential"
+	"github.com/vulpemventures/go-elements/transaction"
+	"github.com/vulpemventures/go-secp256k1-zkp"
+)
+
+type blinder struct {
+	pset             *Pset
+	blindingPrivkeys [][]byte
+	blindingPubkeys  [][]byte
+}
+
+// NewBlinder returns a new instance of blinder, if the passed Pset struct is
+// in a valid form, else an error.
+func NewBlinder(pset *Pset, blindingPrivkeys, blindingPubkeys [][]byte) (
+	*blinder,
+	error,
+) {
+	if err := pset.SanityCheck(); err != nil {
+		return nil, err
+	}
+
+	return &blinder{
+		pset:             pset,
+		blindingPrivkeys: blindingPrivkeys,
+		blindingPubkeys:  blindingPubkeys,
+	}, nil
+}
+
+//BlindOutputs method blinds pset's output
+func (b *blinder) BlindOutputs() error {
+	err := b.validate()
+	if err != nil {
+		return err
+	}
+
+	unblindOutputs, err := b.unblindInputs()
+	if err != nil {
+		return err
+	}
+
+	outputValues := make([]uint64, 0)
+	//TODO check if looping over UnsignedTx's Outputs is correct?
+	for _, output := range b.pset.UnsignedTx.Outputs {
+		var val [9]byte
+		copy(val[:], output.Value)
+		value, err := confidential.ElementsToSatoshiValue(val)
+		if err != nil {
+			return err
+		}
+		outputValues = append(outputValues, value)
+	}
+
+	inputAbfs := make([][]byte, 0)
+	for _, v := range unblindOutputs {
+		inputAbfs = append(inputAbfs, v.AssetBlindingFactor)
+	}
+
+	inputVbfs := make([][]byte, 0)
+	for _, v := range unblindOutputs {
+		inputVbfs = append(inputVbfs, v.ValueBlindingFactor)
+	}
+
+	inputAgs := make([][]byte, 0)
+	for _, v := range unblindOutputs {
+		inputAgs = append(inputAgs, v.Asset)
+	}
+
+	inputValues := make([]uint64, 0)
+	for _, v := range unblindOutputs {
+		inputValues = append(inputValues, v.Value)
+	}
+
+	outputVbfs, outputAbfs, err := b.generateOutputBlindingFactors(
+		inputValues,
+		outputValues,
+		inputAbfs,
+		inputVbfs,
+	)
+	if err != nil {
+		return err
+	}
+
+	err = b.blindOutputs(
+		outputValues,
+		outputAbfs,
+		outputVbfs,
+		inputAgs,
+		inputAbfs,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (b *blinder) validate() error {
+	for _, input := range b.pset.Inputs {
```

Comment:
we should also check that none of the inputs has a `PartialSignature`.

Reply:
Done